File: Functional Programming/[BSc] Computation Theory/Mooshak Problem #1 - Palindromes/tcA.ml
let b = read_int();; (* base numérica *)
let x = read_int();; (* número introduzido *)
let i = 1;; (* variável auxiliar *)
let y = 0;; (* número invertido *)

let rec palindroma = function (x,b,a,y,i) -> 
						
						if (a mod (10*i))>=(b*i) then print_string "ERROR\n"
						else if x=0 && y=a then print_string "YES\n"
						else if x=0 && y!=a then print_string "NO\n"
						else palindroma(x/10,b,a,(y*10)+(x mod 10),i*10);;
palindroma(x,b,x,y,i);;						

File: Functional Programming/[BSc] Computation Theory/Mooshak Problem #1 - Palindromes/tcA.c
#include <stdio.h>

void palindroma(int x, int b, int a, int y, int i){

	if((a%(10*i))>=(b*i)){

		printf("ERROR\n"); //não é um inteiro na base considerada
		return;
	}

	else if(x==0){

		if(y==a){

			printf("YES\n"); //é um palindroma na base considerada
			return;
		}

		printf("NO\n"); //é um inteiro na base considerada, mas não é um palindroma
		return;
	}

	//construir o número invertido
	y=(y*10)+(x%10); 
	x=(x/10);
	i=(i*10);

	palindroma(x,b,a,y,i);
}

int main(int argc, char *argv[]){

	int b=0, x=0, y=0, i=1; //base dada, número dado, número invertido e variável auxiliar

	scanf("%d",&b);
	scanf("%d",&x);
	
	palindroma(x,b,x,y,i);

	return(0);
}